js/client.js

```diff
@@ -828,6 +828,29 @@
 				} else if (data === 'rename') {
 					this.renameRoom(roomid, errormessage);
 				} else if (data !== 'namepending') {
+					if (errormessage) {
+						if (data === 'nonexistent' && Config.server.id && roomid.slice(0, 7) === 'battle-') {
+							var replayid = roomid.slice(7);
+							if (Config.server.id !== 'showdown') replayid = Config.server.id + '-' + replayid;
+							var replayLink = 'https://replay.pokemonshowdown.com/' + replayid;
+							$.ajax(replayLink + '.json', {dataType: 'json'}).done(function (replay) {
+								if (replay) {
+									var title = Tools.escapeHTML(replay.p1) + ' vs. ' + Tools.escapeHTML(replay.p2);
+									app.receive('>battle-' + replayid + '\n|init|battle\n|title|' + title + '\n' + replay.log);
+									app.receive('>battle-' + replayid + '\n|expire|<a href=' + replayLink + ' target="_blank">Open replay in new tab</a>');
+								} else {
+									errormessage += '\n\nResponse received, but no data.';
+									app.addPopupMessage(errormessage);
+								}
+								isdeinit = false;
+							}).fail(function () {
+								errormessage += "\n\nThe battle you're looking for has expired. Battles expire after 15 minutes of inactivity unless they're saved.\nIn the future, remember to click \"Save replay\" to save a replay permanently.";
+								app.addPopupMessage(errormessage);
+							});
+						} else {
+							this.addPopupMessage(errormessage);
+						}
```

Comment:
I would manually `return` after this line, and then manually call `unjoinRoom` in the failure handlers.